pytorch_lightning/trainer/trainer.py

```diff
@@ -1494,6 +1500,42 @@ def call_hook(
 
         return output
 
+    def _profile_and_call(self, obj: object, hook_name: str, *args: Any, **kwargs: Any) -> Optional[Any]:
+        fn = getattr(obj, hook_name)
+        with self.profiler.profile(f"{obj.__class__.__name__}.{hook_name}"):
+            return fn(*args, **kwargs)
+
+    def _call_hook(
```

Comment:
why add a new method instead of migrating `call_hook` in place?

Reply:
do you see a major inconvenience if we break this up like:
`_call_lightning_module_hook`
`_call_callback_hook`
`_call_strategy_hook` (most dubious if we need this)

each of which is profiled accordingly. and then `call_hook` can call these in order

Reply:
> why add a new method instead of migrating call_hook in place?

this is just temporary, the plan is to delete `call_hook`

> and then call_hook can call these in order

I thought one of the main points of this refactor was so that we have the flexibility to call the hooks in any order? i.e. now we do callback hooks first and then LM hooks, and then strategy hooks

> (most dubious if we need this)

I agree, can we simplify things by only having hooks for LM and callbacks?

Reply:
It could still make sense to break it up into:
```
_call_lightning_module_hook
_call_callback_hook
_call_strategy_hook
```

but rather than call them sequentially check which obj is being passed in `_call_hook` and then call the correct function?